core/src/main/java/org/infinispan/expiration/impl/TxClusterExpirationManager.java

```diff
@@ -12,11 +19,46 @@
 
    @Inject
    protected KeyPartitioner partitioner;
+   @Inject @ComponentName(KnownComponentNames.PERSISTENCE_EXECUTOR)
+   protected ExecutorService blockingExecutor;
+
+   private boolean optimisticTransaction;
+
+   @Override
+   public void start() {
+      super.start();
+
+      optimisticTransaction = configuration.transaction().lockingMode() == LockingMode.OPTIMISTIC;
+   }
 
    @Override
    CompletableFuture<Boolean> handleMaxIdleExpireEntry(K key, V value, long maxIdle, boolean skipLocking) {
       // We cannot remove an entry in a tx, due to possible dead lock - thus we just return
       // if the entry is expired or not
-      return retrieveLastAccess(key, value, partitioner.getSegment(key)).thenApply(hasMaxIdle -> hasMaxIdle == null);
+      return retrieveLastAccess(key, value, partitioner.getSegment(key)).thenApply(Objects::isNull);
+   }
+
+   @Override
+   CompletableFuture<Void> removeLifespan(AdvancedCache<K, V> cacheToUse, K key, V value, long lifespan) {
+      // Transactional is still blocking - to be removed later
+      return CompletableFuture.supplyAsync(() -> super.removeLifespan(cacheToUse, key, value, lifespan), blockingExecutor)
+            .thenCompose(Function.identity());
+   }
+
+   @Override
+   CompletableFuture<Boolean> removeMaxIdle(AdvancedCache<K, V> cacheToUse, K key, V value) {
+      // Transactional is still blocking - to be removed later
+      return CompletableFuture.supplyAsync(() -> cacheToUse.removeMaxIdleExpired(key, value), blockingExecutor)
+                  .thenCompose(Function.identity());
+   }
+
+   @Override
+   boolean waitOnLifespanExpiration(boolean hasLock) {
+      // We always have to wait when using optimistic transactions. A scenario where a single writer may call get
+      // and then put may get write skew problems otherwise. The problem is that the get when it sees the expired
+      // entry will fire the remove expired. If we don't wait on this we may end up running the put command at the
+      // same time and get a write skew issue in either caller, which would be very confusing when there is only a
+      // single user of the cache
```

Comment:
Sorry, but I just see a wall of text here... Could you split into shorter sentences, ideally one sentence per line?

Reply:
Sure, I can reformat.